ntoskrnl/ke/i386/traphdlr.c

```diff
@@ -1114,12 +1173,22 @@ KiTrap0DHandler(IN PKTRAP_FRAME TrapFrame)
     /*
      * Check for a fault during checking of the user instruction.
      *
-     * Note that the SEH handler will catch invalid EIP, but we could be dealing
-     * with an invalid CS, which will generate another GPF instead.
-     *
+     * Note that the SEH handler will catch invalid EIP, but we could be
+     * dealing with an invalid CS, which will generate another GPF instead.
      */
-    if (((PVOID)TrapFrame->Eip >= (PVOID)KiTrap0DHandler) &&
-        ((PVOID)TrapFrame->Eip < (PVOID)KiTrap0DHandler))
+    // FIXME BUG! Nothing ensures that the C compiler correctly orders the C trap handlers!!
+    // Bug fix PR #1996
+#if 0
+    // I managed to have the trap handlers sorted as follows:
+    // ... KiTrap(X)() KiTrap(X)Handler() KiTrap(X+1)() KiTrap(X+1)Handler() ...
+    //
+    if (((ULONG_PTR)TrapFrame->Eip >= (ULONG_PTR)KiTrap0D/*Handler*/) &&
+        ((ULONG_PTR)TrapFrame->Eip <  (ULONG_PTR)KiTrap0E/*Handler*/))
+#else
+    // Using (sub)section delimiters...
+    if (((ULONG_PTR)TrapFrame->Eip >= (ULONG_PTR)&_KiTrap0D_start) &&
+        ((ULONG_PTR)TrapFrame->Eip <  (ULONG_PTR)&_KiTrap0D_end  ))
+#endif
```

Comment:
@tkreuzer : It seems that here we really need to somehow check a range of Eip ... (contrary to the kdbg case where we could do something different to see whether we run in a TSS). So what do you think about using either of these two blocks, together with the pragma section stuff I introduced?

Reply:
Would it be possible/easier to use a `static` variable instead, which value would be set/unset around the relevant (whole/part of function) code block?

Reply:
Here's my suggestions to avoid the section stuff:
Solution 1:
- Take ```CallerEip = *(ULONG*)TrapFrame->Rbp;```
- Check whether CallerEip is within KiTrap0D (not Handler!)
This is still a hack, but much simpler than putting everything in sections.

Solution 2 (preferred!):
- Wrap offending code in KeTrap0DHandler (the code that scans user mode code) as well as in KiSystemCallHandler (the code that copies the parameters) in SEH
- Remove this entire block
- Replace the KiSystemFatalException at the bottom with KiDispatchException
We need the SEH anyway, but that would be handled by the page fault handler.
I think the code is unneccessary anyway, since the handler itself cannot create a #GP. In case this was not true, we would be dispatching an exception, handle that and then dispatch the fault back to user mode. If the fault was from somewhere else, we would raise an exception instead of BSODing directly. So the worst effect would be that we allow kernel mode code to catch something that we didn't explicitly handle here with SEH, while it would usually BSOD. I think we can live with that. :)

Reply:
I will implement solution 1, for the moment, but I would really be interested to know what Alex thinks about Solution 2.

Reply:
(Of course, solution 1 would supersede current patch.)